sharding/review/casper_v2.1.md

```diff
@@ -0,0 +1,362 @@
+* **Committee** - a (pseudo-) randomly sampled subset of the active validator set. When a committee is referred to collectively, as in "this committee attests to X", this is assumed to mean "some subset of that committee that contains enough validators that the protocol recognizes it as representing the committee".
+* **Proposer** - the validator that creates a block
+* **Attester** - a validator that is part of a committee that needs to sign off on a block.
+* **Beacon chain** - the central PoS chain that is the base of the sharding system.
+* **Shard chain** - one of the chains on which transactions take place and account data is stored.
+* **Cross-link** - a set of signatures from a committee attesting to a block in a shard chain, which can be included into the beacon chain. Cross-links are the main means by which the beacon chain "learns about" the updated state of shard chains.
+* **Epoch** - a period of 64 blocks.
+* **Finalized**, **justified** - see Casper FFG finalization here: https://arxiv.org/abs/1710.09437
+
+### Constants
+
+* **SHARD_COUNT** - a constant referring to the number of shards. Currently set to 1024.
+* **ETH_SUPPLY_CAP** - self-explanatory. Currently set to 2<sup>27</sup> ~= 134 million.
+* **DEPOSIT_SIZE** - 32 ETH
+* **MAX_VALIDATOR_COUNT** - `ETH_SUPPLY_CAP / DEPOSIT_SIZE = 4194304`
+* **EPOCH_LENGTH** - 64 blocks
+* **SLOT_DURATION** - 8 seconds
+* **MIN_COMMITTEE_SIZE** - 128
+* **END_EPOCH_GRACE_PERIOD** - 8 blocks
+
+### Main chain changes
+
+This PoS/sharding proposal can be implemented separately from the existing PoW main chain. Only two changes to the PoW main chain are required (and the second one is technically not strictly necessary).
+
+* On the PoW main chain a contract is added; this contract allows you to deposit `DEPOSIT_SIZE` ETH; the `deposit` function also takes as arguments (i) `pubkey` (bytes), (ii) `withdrawal_shard_id` (int), (iii)  `withdrawal_addr` (address) and (iv) `randao_commitment` (bytes32)
+* PoW Main chain clients will implement a method, `prioritize(block_hash, value)`. If the block is available and has been verified, this method sets its score to the given value, and recursively adjusts the scores of all descendants. This allows the PoS beacon chain's finality gadget to also implicitly finalize main chain blocks.
+
+### Beacon chain
+
+The beacon chain is the "main chain" of the PoS system. The beacon chain's main responsibilities are:
+
+* Store and maintain the set of active, queued and exited validators
+* Process cross-links (see above) 
+* Process its own block-by-block consensus, as well as the FFG finality gadget
+
+Here are the fields that go into every beacon chain block:
+
+```python
+fields = {
+    # Hash of the parent block
+    'parent_hash': 'hash32',
+    # Slot number (for the PoS mechanism)
+    'slot_number': 'int64',
+    # Randao commitment reveal
+    'randao_reveal': 'hash32',
+    # Attestation votes
+    'attestation_votes': [AttestationVote],
```

Comment:
`AttestationVote` is not defined.

Reply:
Fixed (changed AggregateVote to AttestationVote later)